diff --git a/python/sglang/srt/configs/model_config.py b/python/sglang/srt/configs/model_config.py
index fc549d5a5..6149ad265 100644
--- a/python/sglang/srt/configs/model_config.py
+++ b/python/sglang/srt/configs/model_config.py
@@ -129,6 +129,12 @@ class ModelConfig:
         ):
             self.hf_config.architectures[0] = "DeepseekV3ForCausalLMNextN"
 
+        if (
+            is_draft_model
+            and self.hf_config.architectures[0] == "LongcatFlashForCausalLM"
+        ):
+            self.hf_config.architectures[0] = "LongcatFlashForCausalLMNextN"
+
         if is_draft_model and self.hf_config.architectures[0] == "Glm4MoeForCausalLM":
             self.hf_config.architectures[0] = "Glm4MoeForCausalLMNextN"
 
@@ -221,6 +227,13 @@ class ModelConfig:
             self.qk_nope_head_dim = self.hf_config.qk_nope_head_dim
             self.qk_rope_head_dim = self.hf_config.qk_rope_head_dim
             self.v_head_dim = self.hf_config.v_head_dim
+        elif "LongcatFlashForCausalLMNextN" in self.hf_config.architectures:
+            self.head_dim = 256
+            self.attention_arch = AttentionArch.MLA
+            self.kv_lora_rank = self.hf_config.kv_lora_rank
+            self.qk_nope_head_dim = self.hf_config.qk_nope_head_dim
+            self.qk_rope_head_dim = self.hf_config.qk_rope_head_dim
+            self.v_head_dim = self.hf_config.v_head_dim
         elif "MiniCPM3ForCausalLM" in self.hf_config.architectures:
             self.head_dim = 128
             self.attention_arch = AttentionArch.MLA
@@ -277,6 +290,8 @@ class ModelConfig:
         self.num_attention_layers = self.num_hidden_layers
         if "LongcatFlashForCausalLM" in self.hf_config.architectures:
             self.num_attention_layers = self.num_hidden_layers * 2
+        if "LongcatFlashForCausalLMNextN" in self.hf_config.architectures:
+            self.num_attention_layers = self.num_hidden_layers * 2
         self.num_nextn_predict_layers = getattr(
             self.hf_text_config, "num_nextn_predict_layers", None
         )
diff --git a/python/sglang/srt/model_loader/loader.py b/python/sglang/srt/model_loader/loader.py
index 2e2f71078..96462c63e 100644
--- a/python/sglang/srt/model_loader/loader.py
+++ b/python/sglang/srt/model_loader/loader.py
@@ -577,7 +577,7 @@ class DummyModelLoader(BaseModelLoader):
             if hasattr(model, "post_load_weights"):
                 if (
                     model_config.hf_config.architectures[0]
-                    == "DeepseekV3ForCausalLMNextN"
+                    in ["DeepseekV3ForCausalLMNextN","LongcatFlashForCausalLMNextN"]
                 ):
                     model.post_load_weights(is_nextn=True)
                 else:
diff --git a/python/sglang/srt/models/registry.py b/python/sglang/srt/models/registry.py
index 04bf6caa6..ca2f0d571 100644
--- a/python/sglang/srt/models/registry.py
+++ b/python/sglang/srt/models/registry.py
@@ -111,4 +111,6 @@ module = importlib.import_module("omni.adaptors.sglang.models.deepseek.deepseek_
 tmp_.models["DeepseekV3ForCausalLMNextN"] = module.EntryClass
 module = importlib.import_module("omni.adaptors.sglang.models.longcat.longcat_flash")
 tmp_.models["LongcatFlashForCausalLM"] = module.EntryClass
+module = importlib.import_module("omni.adaptors.sglang.models.longcat.longcat_mtp")
+tmp_.models["LongcatFlashForCausalLMNextN"] = module.EntryClass
 ModelRegistry = tmp_
