diff --git a/vllm/v1/core/sched/utils.py b/vllm/v1/core/sched/utils.py
index 74952eb6b..6154ac558 100644
--- a/vllm/v1/core/sched/utils.py
+++ b/vllm/v1/core/sched/utils.py
@@ -29,16 +29,23 @@ THINK_START_TOKEN_ID: list[int] = [
     int(x.strip()) for x in os.environ.get("THINK_START_TOKEN_ID", "0").split(",") if x.strip().isdigit()
 ]
 THINK_END_TOKEN_ID: int = int(os.environ.get("THINK_END_TOKEN_ID", "0"))
+USE_GPT_OSS = os.environ.get("USE_GPT_OSS", "0") == "1"
 
 def add_token_and_check_stop(token_ids: list[int], request: Request, max_model_len: int):
-    global ENABLE_MAX_TOKENS_EXCLUDE_REASONING, ENABLE_REASONING_MAX_TOKENS, REASONING_MAX_TOKENS, THINK_START_TOKEN_ID, THINK_END_TOKEN_ID
-    if ENABLE_REASONING_MAX_TOKENS and not request.is_cot_end() and request.is_cot_started() and request.num_output_tokens >= REASONING_MAX_TOKENS:
-        request.status = RequestStatus.FINISHED_LENGTH_CAPPED
-        return True, 1
-    
+    global ENABLE_MAX_TOKENS_EXCLUDE_REASONING, ENABLE_REASONING_MAX_TOKENS, REASONING_MAX_TOKENS, THINK_START_TOKEN_ID, THINK_END_TOKEN_ID, USE_GPT_OSS
+    if USE_GPT_OSS:
+        if ENABLE_REASONING_MAX_TOKENS and not request.is_cot_end() and request.is_cot_started() and request.num_output_tokens >= REASONING_MAX_TOKENS:
+            request.status = RequestStatus.FINISHED_LENGTH_CAPPED
+            return True, 1
+    else:
+        if ENABLE_REASONING_MAX_TOKENS and not request.is_cot_end() and request.num_output_tokens >= REASONING_MAX_TOKENS:
+            request.status = RequestStatus.FINISHED_LENGTH_CAPPED
+            return True, 1
     is_able_to_add_max_tokens = request.num_prompt_tokens + request.max_tokens < max_model_len
     if ENABLE_MAX_TOKENS_EXCLUDE_REASONING and is_able_to_add_max_tokens and not request.is_cot_end():
-        is_cot_started = request.is_cot_started()
+        is_cot_started = True
+        if USE_GPT_OSS:
+            is_cot_started = request.is_cot_started()
         # If the start of COT is recognized in this turn and its THINK_START_TOKEN_ID is long in size, then the
         # tokens consumed by THINK_START_TOKEN_ID in previous turns need to be returned
         if not is_cot_started:
