diff --git a/vllm/v1/core/kv_cache_manager.py b/vllm/v1/core/kv_cache_manager.py
index 04dc331a1..d0d3037b6 100644
--- a/vllm/v1/core/kv_cache_manager.py
+++ b/vllm/v1/core/kv_cache_manager.py
@@ -15,6 +15,7 @@ from vllm.v1.core.single_type_kv_cache_manager import (
 from vllm.v1.kv_cache_interface import KVCacheConfig
 from vllm.v1.metrics.stats import PrefixCacheStats
 from vllm.v1.request import Request, RequestStatus
+from vllm.config import SpeculativeConfig
 
 logger = init_logger(__name__)
 
@@ -62,6 +63,7 @@ class KVCacheManager:
         log_stats: bool = False,
         enable_kv_cache_events: bool = False,
         combine_block_num: int = 1,
+        speculative_config: Optional[SpeculativeConfig] = None,
     ) -> None:
         if len(kv_cache_config.kv_cache_groups) > 1 and \
             len(set(type(group.kv_cache_spec) for group in kv_cache_config.kv_cache_groups)) != 1:
@@ -83,6 +85,7 @@ class KVCacheManager:
 
         self.single_type_manager = get_manager_for_kv_cache_spec(
             kv_cache_spec=kv_cache_spec,
+            speculative_config=speculative_config,
             block_pool=self.block_pool,
             use_eagle=self.use_eagle,
             num_kv_cache_groups=1,
diff --git a/vllm/v1/core/sched/scheduler.py b/vllm/v1/core/sched/scheduler.py
index 9ebac704d..fafd08045 100644
--- a/vllm/v1/core/sched/scheduler.py
+++ b/vllm/v1/core/sched/scheduler.py
@@ -183,6 +183,7 @@ class Scheduler(SchedulerInterface):
         if self.combine_block_num > 1:
             self.kv_cache_manager = KVCacheManager(
                 kv_cache_config=kv_cache_config,
+                speculative_config=speculative_config,
                 max_model_len=self.max_model_len,
                 enable_caching=self.cache_config.enable_prefix_caching,
                 caching_hash_algo=self.cache_config.prefix_caching_hash_algo,
@@ -194,6 +195,7 @@ class Scheduler(SchedulerInterface):
         else:
             self.kv_cache_manager = KVCacheManager(
                 kv_cache_config=kv_cache_config,
+                speculative_config=speculative_config,
                 max_model_len=self.max_model_len,
                 enable_caching=self.cache_config.enable_prefix_caching,
                 caching_hash_algo=self.cache_config.prefix_caching_hash_algo,
@@ -218,6 +220,7 @@ class Scheduler(SchedulerInterface):
 
                 cpu_kv_cache_manager = KVCacheManager(
                     kv_cache_config=self.cpu_kv_cache_config,
+                    speculative_config=speculative_config,
                     max_model_len=self.max_model_len,
                     enable_caching=False,
                 )
diff --git a/vllm/v1/core/single_type_kv_cache_manager.py b/vllm/v1/core/single_type_kv_cache_manager.py
index cd24fb4bb..a7697bdee 100644
--- a/vllm/v1/core/single_type_kv_cache_manager.py
+++ b/vllm/v1/core/single_type_kv_cache_manager.py
@@ -2,6 +2,7 @@
 from abc import ABC, abstractmethod
 from collections import defaultdict
 from typing import Callable
+from typing import Optional
 
 from vllm.utils import cdiv
 from vllm.v1.core.block_pool import BlockPool
@@ -9,6 +10,7 @@ from vllm.v1.core.kv_cache_utils import BlockHashType, KVCacheBlock
 from vllm.v1.kv_cache_interface import (FullAttentionSpec, KVCacheSpec,
                                         SlidingWindowSpec)
 from vllm.v1.request import Request
+from vllm.config import SpeculativeConfig
 
 
 class SingleTypeKVCacheManager(ABC):
@@ -25,6 +27,7 @@ class SingleTypeKVCacheManager(ABC):
         num_kv_cache_groups: int,
         caching_hash_fn: Callable,
         combine_block_num: int = 1,
+        speculative_config: Optional[SpeculativeConfig] = None,
     ) -> None:
         """
         Initializes the SpecializedManager.
@@ -59,6 +62,8 @@ class SingleTypeKVCacheManager(ABC):
         self.num_kv_cache_groups = num_kv_cache_groups
         self.caching_hash_fn = caching_hash_fn
         self.combine_block_num = combine_block_num
+        if self.use_eagle and speculative_config is not None:
+            self.num_spec_tokens = speculative_config.num_speculative_tokens
 
     def get_num_blocks_to_allocate(
             self, request_id: str, num_tokens: int,
@@ -248,8 +253,9 @@ class FullAttentionManager(SingleTypeKVCacheManager):
                 computed_blocks.append(cached_block)
             else:
                 break
-        if self.use_eagle and len(computed_blocks) > 0:
-            computed_blocks.pop()
+        if self.use_eagle and len(computed_blocks) > 0 and len(computed_blocks) == max_num_blocks:
+            if self.num_spec_tokens is None or self.num_spec_tokens == 0 or max_length % self.block_size < self.num_spec_tokens:
+                computed_blocks.pop()
         return computed_blocks
 
     def remove_skipped_blocks(self, request_id: str,
@@ -319,8 +325,9 @@ class SlidingWindowManager(SingleTypeKVCacheManager):
             # The first `num_contiguous_blocks` is a cache hit even if
             # `num_contiguous_blocks < sliding_window_contiguous_blocks`.
             del computed_blocks[num_contiguous_blocks:]
-        if self.use_eagle and len(computed_blocks) > 0:
-            computed_blocks.pop()
+        if self.use_eagle and len(computed_blocks) > 0 and len(computed_blocks) == max_num_blocks:
+            if self.num_spec_tokens is None or self.num_spec_tokens == 0 or max_length % self.block_size < self.num_spec_tokens:
+                computed_blocks.pop()
         return computed_blocks
 
     def remove_skipped_blocks(self, request_id: str,
