diff --git a/vllm/v1/core/sched/scheduler.py b/vllm/v1/core/sched/scheduler.py
index 1f01c1e0a..a58ab11c5 100755
--- a/vllm/v1/core/sched/scheduler.py
+++ b/vllm/v1/core/sched/scheduler.py
@@ -91,6 +91,7 @@ class Scheduler(SchedulerInterface):
 
         # Scheduling constraints.
         self.max_num_running_reqs = self.scheduler_config.max_num_seqs
+        self.max_num_scheduled_reqs = self.scheduler_config.max_num_seqs
         self.max_num_scheduled_tokens = \
             self.scheduler_config.max_num_batched_tokens
         self.max_model_len = self.scheduler_config.max_model_len
@@ -264,6 +265,8 @@ class Scheduler(SchedulerInterface):
         blocks_to_swap_out: list[list[tuple[int, int]]] = [[] for _ in range(num_groups)]
         blocks_to_swap_in: list[list[tuple[int, int]]] = [[] for _ in range(num_groups)]
 
+        num_scheduled_reqs = 0
+
         # For logging.
         scheduled_timestamp = time.monotonic()
 
@@ -374,6 +377,7 @@ class Scheduler(SchedulerInterface):
 
             # Schedule the request.
             scheduled_running_reqs.append(request)
+            num_scheduled_reqs += 1
             if request.use_structured_output:
                 # PERF: in case of chunked prefill,
                 # request might not include any new tokens.
@@ -434,9 +438,10 @@ class Scheduler(SchedulerInterface):
         skipped_waiting_requests = create_request_queue(self.policy)
 
         # Next, schedule the WAITING requests.
+        pp_size = self.vllm_config.parallel_config.pipeline_parallel_size
         if not preempted_reqs:
             while self.waiting and token_budget > 0:
-                if len(self.running) == self.max_num_running_reqs:
+                if (pp_size == 1 and len(self.running) == self.max_num_running_reqs) or num_scheduled_reqs == self.max_num_scheduled_reqs:
                     break
 
                 request = self.waiting.peek_request()
@@ -596,6 +601,7 @@ class Scheduler(SchedulerInterface):
                 else:
                     raise RuntimeError(
                         f"Invalid request status: {request.status}")
+                num_scheduled_reqs += 1
 
                 if self.lora_config and request.lora_request:
                     scheduled_loras.add(request.lora_request.lora_int_id)
@@ -625,12 +631,15 @@ class Scheduler(SchedulerInterface):
         total_num_scheduled_tokens = sum(num_scheduled_tokens.values())
         assert total_num_scheduled_tokens <= self.max_num_scheduled_tokens
         assert token_budget >= 0
-        assert len(self.running) <= self.max_num_running_reqs
+        if pp_size == 1:
+            assert len(self.running) <= self.max_num_running_reqs
         # Since some requests in the RUNNING queue may not be scheduled in
         # this step, the total number of scheduled requests can be smaller than
         # len(self.running).
         assert (len(scheduled_new_reqs) + len(scheduled_resumed_reqs) +
-                len(scheduled_running_reqs) <= len(self.running))
+                len(scheduled_running_reqs) == num_scheduled_reqs)
+        assert num_scheduled_reqs <= len(self.running)
+        assert num_scheduled_reqs <= self.max_num_scheduled_reqs
 
         # Get the longest common prefix among all requests in the running queue.
         # This can be potentially used for cascade attention.
