diff --git a/vllm/entrypoints/openai/serving_engine.py b/vllm/entrypoints/openai/serving_engine.py
index d4e840ccc..3239f55b6 100644
--- a/vllm/entrypoints/openai/serving_engine.py
+++ b/vllm/entrypoints/openai/serving_engine.py
@@ -412,7 +412,7 @@ class OpenAIServing:
                     f"{token_num} tokens in the messages, "
                     f"Please reduce the length of the messages.")
         elif token_num + max_tokens > max_model_len:
-            raise ValueError(
+            logger.warning(
                 f"This model's maximum context length is "
                 f"{max_model_len} tokens. However, you requested "
                 f"{max_tokens + token_num} tokens "
@@ -798,7 +798,7 @@ class OpenAIServing:
                     f"{token_num} tokens in the messages, "
                     f"Please reduce the length of the messages.")
         elif token_num + max_tokens > self.max_model_len:
-            raise ValueError(
+            logger.warning(
                 f"This model's maximum context length is "
                 f"{self.max_model_len} tokens. However, you requested "
                 f"{max_tokens + token_num} tokens "
