diff --git a/vllm/config.py b/vllm/config.py
index 5c6708c7d..506cf9268 100644
--- a/vllm/config.py
+++ b/vllm/config.py
@@ -1322,6 +1322,7 @@ class ModelConfig:
             "top_p",
             "min_p",
             "max_new_tokens",
+            "top_n_sigma",
         ]
         if any(p in config for p in available_params):
             diff_sampling_param = {
diff --git a/vllm/entrypoints/openai/protocol.py b/vllm/entrypoints/openai/protocol.py
index a92965dbe..b02fdedd3 100644
--- a/vllm/entrypoints/openai/protocol.py
+++ b/vllm/entrypoints/openai/protocol.py
@@ -269,6 +269,7 @@ class ChatCompletionRequest(OpenAIBaseModel):
     spaces_between_special_tokens: bool = True
     truncate_prompt_tokens: Optional[Annotated[int, Field(ge=1)]] = None
     prompt_logprobs: Optional[int] = None
+    top_n_sigma: Optional[float] = None
     # --8<-- [end:chat-completion-sampling-params]
 
     # --8<-- [start:chat-completion-extra-params]
@@ -424,6 +425,7 @@ class ChatCompletionRequest(OpenAIBaseModel):
         "top_p": 1.0,
         "top_k": 0,
         "min_p": 0.0,
+        "top_n_sigma": -1,
     }
 
     def to_beam_search_params(
@@ -493,6 +495,9 @@ class ChatCompletionRequest(OpenAIBaseModel):
         if (min_p := self.min_p) is None:
             min_p = default_sampling_params.get(
                 "min_p", self._DEFAULT_SAMPLING_PARAMS["min_p"])
+        if (top_n_sigma := self.top_n_sigma) is None:
+            top_n_sigma = default_sampling_params.get(
+                "top_n_sigma", self._DEFAULT_SAMPLING_PARAMS["top_n_sigma"])
 
         prompt_logprobs = self.prompt_logprobs
         if prompt_logprobs is None and self.echo:
@@ -527,6 +532,10 @@ class ChatCompletionRequest(OpenAIBaseModel):
         if self.kv_transfer_params:
             # Pass in kv_transfer_params via extra_args
             extra_args["kv_transfer_params"] = self.kv_transfer_params
+
+        if top_n_sigma:
+            extra_args["top_n_sigma"] = top_n_sigma
+
         return SamplingParams.from_optional(
             n=self.n,
             best_of=self.best_of,
@@ -789,6 +798,7 @@ class CompletionRequest(OpenAIBaseModel):
     truncate_prompt_tokens: Optional[Annotated[int, Field(ge=1)]] = None
     allowed_token_ids: Optional[list[int]] = None
     prompt_logprobs: Optional[int] = None
+    top_n_sigma: Optional[float] = None
     # --8<-- [end:completion-sampling-params]
 
     # --8<-- [start:completion-extra-params]
@@ -882,6 +892,7 @@ class CompletionRequest(OpenAIBaseModel):
         "top_p": 1.0,
         "top_k": 0,
         "min_p": 0.0,
+        "top_n_sigma": -1,
     }
 
     def to_beam_search_params(
@@ -948,6 +959,9 @@ class CompletionRequest(OpenAIBaseModel):
         if (min_p := self.min_p) is None:
             min_p = default_sampling_params.get(
                 "min_p", self._DEFAULT_SAMPLING_PARAMS["min_p"])
+        if (top_n_sigma := self.top_n_sigma) is None:
+            top_n_sigma = default_sampling_params.get(
+                "top_n_sigma", self._DEFAULT_SAMPLING_PARAMS["top_n_sigma"])
 
         prompt_logprobs = self.prompt_logprobs
         if prompt_logprobs is None and self.echo:
@@ -974,6 +988,10 @@ class CompletionRequest(OpenAIBaseModel):
         if self.kv_transfer_params:
             # Pass in kv_transfer_params via extra_args
             extra_args["kv_transfer_params"] = self.kv_transfer_params
+        
+        if top_n_sigma :
+            extra_args["top_n_sigma"] = top_n_sigma    
+    
         return SamplingParams.from_optional(
             n=self.n,
             best_of=self.best_of,
diff --git a/vllm/v1/sample/metadata.py b/vllm/v1/sample/metadata.py
index b1816d758..3ade689b7 100644
--- a/vllm/v1/sample/metadata.py
+++ b/vllm/v1/sample/metadata.py
@@ -43,3 +43,5 @@ class SamplingMetadata:
     bad_words_token_ids: dict[int, list[list[int]]]
     
     seq_data: list
+    
+    top_n_sigma: Optional[torch.Tensor]
diff --git a/vllm/v1/sample/sampler.py b/vllm/v1/sample/sampler.py
index 225965b41..55190ed2c 100644
--- a/vllm/v1/sample/sampler.py
+++ b/vllm/v1/sample/sampler.py
@@ -124,6 +124,10 @@ class Sampler(nn.Module):
         if sampling_metadata.min_p is not None:
             logits = self.apply_min_p(logits, sampling_metadata.min_p)
 
+        # Apply top_n_sigma
+        if sampling_metadata.top_n_sigma is not None:
+            logits = self.apply_top_n_sigma(logits, sampling_metadata.top_n_sigma)
+
         # Apply top_k and/or top_p.
         random_sampled = self.topk_topp_sampler(
             logits,
@@ -305,4 +309,19 @@ class Sampler(nn.Module):
         if self.enable_early_stop:
             return self.think_compressor.update_sampled_token_ids(token_ids, sampling_metadata.seq_data)
 
-        return token_ids
\ No newline at end of file
+        return token_ids
+
+
+    def apply_top_n_sigma(self, logits: torch.Tensor, top_n_sigma: torch.Tensor):
+        
+        top_n_sigma = top_n_sigma.unsqueeze(-1)
+        top_n_sigma_mask = top_n_sigma != -1
+        
+        filter_value = -3.4028e38
+        max_vals, _ = logits.max(dim=-1, keepdim=True)
+        std_vals = logits.std(dim=-1, keepdim=True)
+        threshold = max_vals - top_n_sigma * std_vals
+        threshold = torch.where(top_n_sigma_mask, threshold, filter_value)
+        mask = logits < threshold
+        logits = torch.where(mask, filter_value, logits)
+        return logits
\ No newline at end of file
diff --git a/vllm/v1/worker/gpu_input_batch.py b/vllm/v1/worker/gpu_input_batch.py
index b0534ed73..4ccaa00aa 100644
--- a/vllm/v1/worker/gpu_input_batch.py
+++ b/vllm/v1/worker/gpu_input_batch.py
@@ -151,7 +151,20 @@ class InputBatch:
                                             pin_memory=pin_memory)
         self.min_p_cpu = self.min_p_cpu_tensor.numpy()
         self.min_p_reqs: set[str] = set()
-
+        
+        # topnsigma penalty
+        self.top_n_sigma = torch.empty((max_num_reqs, ),
+                                               dtype=torch.float,
+                                               device=device)
+        self.top_n_sigma_cpu_tensor = torch.empty(
+            (max_num_reqs, ),
+            dtype=torch.float,
+            device="cpu",
+            pin_memory=pin_memory)
+        self.top_n_sigma_cpu = \
+            self.top_n_sigma_cpu_tensor.numpy()
+        self.top_n_sigma_reqs: set[str] = set()
+ 
         # Frequency penalty related data structures
         self.frequency_penalties = torch.empty((max_num_reqs, ),
                                                dtype=torch.float,
@@ -316,6 +329,13 @@ class InputBatch:
         if sampling_params.min_tokens:
             self.min_tokens[req_index] = (sampling_params.min_tokens,
                                           sampling_params.all_stop_token_ids)
+        # patch top_n_sigma
+        if sampling_params.extra_args and "top_n_sigma" in sampling_params.extra_args:
+            self.top_n_sigma_cpu[
+                    req_index] = sampling_params.extra_args["top_n_sigma"]
+            self.top_n_sigma_reqs.add(req_id)
+        else:
+            self.top_n_sigma_cpu[req_index] = -1
             
         # patch for thinking budget control for each request
         if sampling_params.extra_args and "thinking_budget" in sampling_params.extra_args and\
@@ -395,6 +415,8 @@ class InputBatch:
         self.num_prompt_logprobs.pop(req_id, None)
         self.in_progress_prompt_logprobs_cpu.pop(req_id, None)
         self.thinking_budget.pop(req_index, None)
+        # patch top_n_sigma
+        self.top_n_sigma_reqs.discard(req_id)
 
         # LoRA
         lora_id = self.request_lora_mapping[req_index]
@@ -448,6 +470,9 @@ class InputBatch:
             
             
         swap_dict_values(self.thinking_budget, i1, i2)
+        # patch top_n_sigma
+        self.top_n_sigma_cpu[i1], self.top_n_sigma_cpu[i2] =\
+            self.top_n_sigma_cpu[i2], self.top_n_sigma_cpu[i1]
 
         # NOTE: the following is unsafe
         # self.token_ids_cpu[i1, ...], self.token_ids_cpu[i2, ...], =\
@@ -531,6 +556,10 @@ class InputBatch:
             thinking_budget = self.thinking_budget.pop(last_req_index, None)
             if thinking_budget is not None:
                 self.thinking_budget[empty_index] = thinking_budget
+            
+            # patch top_n_sigma 
+            self.top_n_sigma_cpu[
+                empty_index] = self.top_n_sigma_cpu[last_req_index]
                 
             generator = self.generators.pop(last_req_index, None)
             if generator is not None:
@@ -595,7 +624,12 @@ class InputBatch:
             prompt_token_ids = self._make_prompt_token_ids_tensor()
         else:
             prompt_token_ids = None
-
+            
+        # patch top_n_sigma
+        if not self.no_top_n_sigma:
+            copy_slice(self.top_n_sigma_cpu_tensor,
+                       self.top_n_sigma, num_reqs)
+ 
         allowed_token_ids_mask: Optional[torch.Tensor] = None
         if not self.no_allowed_token_ids:
             assert self.allowed_token_ids_mask is not None
@@ -625,6 +659,7 @@ class InputBatch:
             allowed_token_ids_mask=allowed_token_ids_mask,
             bad_words_token_ids=self.bad_words_token_ids,
             seq_data=all_seq_data,
+            top_n_sigma=self.top_n_sigma[:num_reqs],
         )
         
     def _handle_early_think_stopping(self, num_reqs):
@@ -733,3 +768,9 @@ class InputBatch:
     @property
     def no_allowed_token_ids(self) -> bool:
         return len(self.has_allowed_token_ids) == 0
+    
+    # patch top_n_sigma
+    @property 
+    def no_top_n_sigma(self) -> bool:
+        return len(self.top_n_sigma_reqs) == 0
+ 
\ No newline at end of file
